tests/cpp/test_multidevice_transformer.cpp

```diff
@@ -295,55 +448,165 @@ std::vector<TensorView*> mlp_backwards(
-  TensorView* matmul0_grad_x = sum(matmul0_grad_x_partial, {0}); // allreduce
-  TensorView* grad_gelu_t = transpose(gelu_grad_f, 1, 2);
-  TensorView* matmul0_grad_w_t = matmul(grad_gelu_t, x);
-  TensorView* matmul0_grad_w = transpose(matmul0_grad_w_t, 1, 2);
-  TensorView* matmul0_grad_b = sum(gelu_grad, {1});
+  auto linear0_grads = linear_backwards(x, w0, gelu_grad, dtype);
 
   // Manaul sharding annotations
   for (auto tv :
-       {x, grad, mask, dropout_grad, matmul1_grad_b, matmul0_grad_x}) {
+       {x,
+        grad,
+        mask,
+        dropout_grad,
+        linear1_grads.grad_b,
+        linear0_grads.grad_x}) {
     tv->setDeviceMesh(mesh);
   }
 
   for (auto tv :
        {w0,
         b0,
         w1,
-        matmul1_grad_w,
+        linear1_grads.grad_x,
+        linear1_grads.grad_w,
         gelu_grad,
-        matmul0_grad_w,
-        matmul0_grad_b}) {
+        linear0_grads.grad_w,
+        linear0_grads.grad_b}) {
     tv->setDeviceMesh(mesh);
     tv->axis(0)->parallelize(ParallelType::DIDx);
   }
 
   std::vector<TensorView*> outputs = {
       dropout_grad,
-      matmul1_grad_w,
-      matmul1_grad_b,
+      linear1_grads.grad_w,
+      linear1_grads.grad_b,
       gelu_grad,
-      matmul0_grad_w,
-      matmul0_grad_b,
-      matmul0_grad_x};
+      linear0_grads.grad_w,
+      linear0_grads.grad_b,
+      linear0_grads.grad_x};
   return outputs;
 }
+
+std::vector<TensorView*> mha_backwards(
+    TensorView* x,
+    TensorView* w0,
+    TensorView* b0,
+    TensorView* w1,
+    TensorView* mask,
+    TensorView* sdpa_output,
+    TensorView* sdpa_log_sumexp,
+    TensorView* sdpa_seed,
+    TensorView* sdpa_offset,
+    TensorView* grad,
+    const DeviceMesh& mesh,
+    DataType dtype) {
+  // Recompute: Linear 1, QKV vectors
+  TensorView* mm = matmul(x, w0);
+  TensorView* proj_bias_bcast = broadcast(b0, {false, true, false});
+  TensorView* qkv1 = add(mm, proj_bias_bcast);
+  // Forming the q,k,v vectors:
+  auto D = w0->axis(0)->extent()->value().as<int64_t>();
+  TensorView* qkv = reshape(qkv1, {D, B * S, 3 * E / D}, {D, B, S, 3 * E / D});
+  std::vector<TensorView*> qkv_reshaped = {};
+  for (auto i : c10::irange(3)) {
+    TensorView* tv_slice =
+        slice(qkv, {0, 0, 0, E / D * i}, {D, B, S, E / D * (i + 1)});
+    TensorView* tv_reshape =
+        reshape(tv_slice, {D, B, S, E / D}, {D, B, S, H / D, E / H});
+    TensorView* tv_trans = transpose(tv_reshape, 2, 3);
+    TensorView* tv_cast = castOp(dtype, tv_trans);
+    qkv_reshaped.push_back(tv_cast);
+    // Explicitly shard qkv before calling SDPA node
+    for (auto tv : {tv_slice, tv_reshape, tv_trans, tv_cast}) {
+      tv->setDeviceMesh(mesh);
+      tv->axis(0)->parallelize(ParallelType::DIDx);
+    }
+  }
+
+  // Backwards
+  constexpr double kScale = 1.0 / (1.0 - kDropoutProb);
+  Val* dscale = IrBuilder::create<Val>(kScale);
+  TensorView* dropout_grad = dropout_backward(grad, mask, dscale);
```

Comment:
What does `d` in `dscale` stand for? If it means, I'd say `dropout_scale` for clarity.